src/Microsoft.ML.Vision/ImageClassificationTrainer.cs

```diff
@@ -937,7 +937,7 @@ private int GetNumSamples(string path)
                 metrics.Train.LearningRate = learningRate;
                 // Update train state.
                 trainstate.CurrentEpoch = epoch;
-                using (var cursor = trainingSet.GetRowCursor(trainingSet.Schema.ToArray(), new Random()))
+                using (var cursor = trainingSet.GetRowCursor(trainingSet.Schema.ToArray(), Host.Rand))
```

Comment:
>trainingSet.GetRowCursor(trainingSet.Schema.ToArray(), Host.Rand)) [](start = 36, length = 66)

You don't need to pass anything, look at the code for ShuffleRowTransformer:

            if (_forceShuffle || _forceShuffleSource)
                _forceShuffleSeed = options.ForceShuffleSeed ?? Host.Rand.NextSigned();

Reply:
While you are correct that if we do not pass anything then it should still be stable, RowShuffling uses host to create a seed to then create a new random, this seems unnecessary, as we would be creating a new object when we can just use an existing random which is more efficient.

---
In reply to: [344297118](https://github.com/dotnet/machinelearning/pull/4446#discussion_r344297118) [](ancestors = 344297118)

Reply:
I discussed offline with @bpstark and we agreed that we should not be passing in Host.Rand object (as I suggested earlier) because it is not thread safe. As I explained, the shuffling transformer code will create a seed from Host.Rand object (that it has access internally) to create an local/internal Rand object and this will ensure stability and also remove concerns around thread safety.